### PR TITLE
Support Web builds in buildFlutterApplication

### LINKED ISSUE #254830: flutter.buildFlutterApplication is limited to linux building.
I wanted to build the web version of FluffyChat.
Currently, `flutter.builFlutterApplication` is designed to handle the linux target, not the web target.
So I needed to override the `buildPhase`, `installPhase` and `postFixup` of `buildFlutterApplication`, but it's actually not possible because the user-declared arguments are overridden by the default ones (see [here](https://github.com/NixOS/nixpkgs/blob/b9ab9a9/pkgs/build-support/flutter/default.nix#L61)).
I'd like to make the contrary.
Instead of this
```nix
(finalAttrs: args // { ... })
```
Do this
```nix
(finalAttrs: { ... } // args)
```
It would allow customizing every part when needed.
Is there any downside I didn't think of ? Is there an easier way to go?
Thanks a lot!

Flutter maintainers: @babariviere @ericdallo @FlafyDev @gilice @hacker1024

### PR DESCRIPTION
**Pull request by hacker1024**
## Description of changes

This PR refactors `buildFlutterApplication` in a modular fashion, allowing it to be more easily extended to support building for other platforms. Web has been implemented, and Android could be done in the future.

FluffyChat has had Web support added as well.

This PR also brings back the `debug` output for Linux builds, which was accidentally removed at some point.

Closes #254830.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

@NixOS/flutter

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by hacker1024** [comment]
Result of `nixpkgs-review pr 281662` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.nbdev</li>
    <li>python312Packages.nbdev.dist</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>discourse</li>
    <li>discourseAllPlugins</li>
  </ul>
</details>
<details>
  <summary>38 packages built:</summary>
  <ul>
    <li>dart-sass</li>
    <li>dart-sass.pubcache</li>
    <li>domine</li>
    <li>domine.pubcache</li>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.debug</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.debug</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat-web</li>
    <li>fluffychat-web.pubcache</li>
    <li>fluffychat.debug</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.debug</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.debug</li>
    <li>localsend.pubcache</li>
    <li>protoc-gen-dart</li>
    <li>protoc-gen-dart.pubcache</li>
    <li>python311Packages.nbdev</li>
    <li>python311Packages.nbdev.dist</li>
    <li>quarto</li>
    <li>quartoMinimal</li>
    <li>rstudio</li>
    <li>rstudio-server</li>
    <li>rstudioServerWrapper</li>
    <li>rstudioWrapper</li>
    <li>shopware-cli</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.debug</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

**Comment by hacker1024** [comment]
Result of `nixpkgs-review pr 281662` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.nbdev</li>
    <li>python312Packages.nbdev.dist</li>
    <li>rstudio</li>
    <li>rstudio-server</li>
    <li>rstudioServerWrapper</li>
    <li>rstudioWrapper</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>discourse</li>
    <li>discourseAllPlugins</li>
  </ul>
</details>
<details>
  <summary>34 packages built:</summary>
  <ul>
    <li>dart-sass</li>
    <li>dart-sass.pubcache</li>
    <li>domine</li>
    <li>domine.pubcache</li>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.debug</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.debug</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat-web</li>
    <li>fluffychat-web.pubcache</li>
    <li>fluffychat.debug</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.debug</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.debug</li>
    <li>localsend.pubcache</li>
    <li>protoc-gen-dart</li>
    <li>protoc-gen-dart.pubcache</li>
    <li>python311Packages.nbdev</li>
    <li>python311Packages.nbdev.dist</li>
    <li>quarto</li>
    <li>quartoMinimal</li>
    <li>shopware-cli</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.debug</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

**Comment by hacker1024** [comment]
`discourse-assets` is already broken.

**Comment by FlafyDev** [comment]
How about changing `flutterHostPlatform` to something like `flutterPlatform`? "hostPlatform" reminds me of Nix's hostPlatform which isn't related

Also I think in several other places in the code already it's regarded as `flutterPlatform` as well

**Comment by hacker1024** [comment]
> How about changing `flutterHostPlatform` to something like `flutterPlatform`? "hostPlatform" reminds me of Nix's hostPlatform which isn't related
> 
> Also I think in several other places in the code already it's regarded as `flutterPlatform` as well

`flutterPlatform` is a little ambiguous, though. Is is the build or runtime platform? The wrapper uses `supportedTargetFlutterPlatforms`, so what about `targetFlutterPlatform`?

**Comment by FlafyDev** [comment]
> `flutterPlatform` is a little ambiguous, though. Is is the build or runtime platform? The wrapper uses `supportedTargetFlutterPlatforms`, so what about `targetFlutterPlatform`?

Yeah `targetFlutterPlatform` is good

File: pkgs/applications/networking/instant-messengers/fluffychat/default.nix [hunk 1]
```diff
@@ -59,15 +78,21 @@ flutter313.buildFlutterApplication rec {
...
-    maintainers = with maintainers; [ mkg20001 gilice ];
-    platforms = [ "x86_64-linux" "aarch64-linux" ];
-    sourceProvenance = [ sourceTypes.fromSource ];
-  };
-}
+} // lib.optionalAttrs (targetFlutterPlatform == "web") {
+  prePatch =
+    # https://github.com/krille-chan/fluffychat/blob/v1.17.1/scripts/prepare-web.sh
+    let
+      # Use Olm 1.3.2, the oldest version, for FluffyChat 1.14.1 which depends on olm_flutter 1.2.0.
```
Comment (mkg20001) [inline pkgs/applications/networking/instant-messengers/fluffychat/default.nix:85]: could fluffychat be updated in this pr aswell? latest is v1.17.1 https://github.com/krille-chan/fluffychat/releases/tag/v1.17.1

Comment (mkg20001) [reply pkgs/applications/networking/instant-messengers/fluffychat/default.nix:85]: yes

**Review by FlafyDev — APPROVED** [review]

**Review by mkg20001 — APPROVED** [review]

**Comment by zh4ngx** [comment]
Thanks for this! Any tips on how to get Android builds working on `buildFlutterApplication`?

**Pull request merged**